include/private/gc_priv.h

```diff
@@ -2116,11 +2116,14 @@ GC_INNER void GC_with_callee_saves_pushed(void (*fn)(ptr_t, void *),
   /* Load value, check tag and permissions of the target memory.   */
 # define LOAD_WORD_OR_CONTINUE(v, p) \
         { \
-            v = *(void **)(p);                                            \
-            size_t has_rwx = cheri_perms_get(v) & (CHERI_PERM_LOAD        \
-                                                   | CHERI_PERM_STORE     \
-                                                   | CHERI_PERM_EXECUTE); \
-            if ((cheri_tag_get(v) == 0) || !has_rwx) continue;            \
+            v = *(void **)(p);                                               \
+            size_t in_bounds = (v >= (word)cheri_base_get(v))                \
+                                && (v < (word)cheri_base_get(v)              \
+                                        + (word)cheri_length_get(v));        \
```

Comment:
I *wonder*: if a naughty person created an invalid capability with a very high base and a very long length, could this `+` overflow? If so, we'd be in undefined behaviour (I think...) and I wonder what would happen?

Reply:
In the case of the GC, **if** this was a capability with a valid tag **and**  the only reference, it would end up causing the GC to collect the memory it points to in the heap.

Reply:
In the case where the tag is valid, it would cause GC to put this on the mark stack and cause a runtime **SIGPROT**  when dereferencing the capability later in the marking stage.

Reply:
What about the *invalid* case?

Reply:
The *invalid* case should be protected by the `!cheri_tag_get()`  clause in the `if` condition

Reply:
This is now fixed in https://github.com/capablevms/bdwgc/pull/19/commits/07b586b7ec011b520062977bd8e863a51061e711, right?

Reply:
> This is now fixed in [07b586b](https://github.com/capablevms/bdwgc/commit/07b586b7ec011b520062977bd8e863a51061e711), right?

Yes the same change to check the `cheri_tag_get()`  before extracting bounds  is fine I think.  Done in [07b586b](https://github.com/capablevms/bdwgc/pull/19/commits/07b586b7ec011b520062977bd8e863a51061e711)